paasta_tools/autoscaling_lib.py

```diff
@@ -12,97 +12,304 @@
     return outer
 
 
-def get_autoscaling_method(function_name):
-    return _autoscaling_methods[function_name]
+def get_autoscaling_ingester(name):
+    return _autoscaling_ingesters[name]
 
 
-def compose_autoscaling_zookeeper_root(service, instance):
-    return '/autoscaling/%s/%s' % (service, instance)
+def get_autoscaling_decider(name):
+    return _autoscaling_deciders[name]
 
 
-def set_instances_for_marathon_service(service, instance, instance_count, soa_dir=DEFAULT_SOA_DIR):
-    zookeeper_path = '%s/instances' % compose_autoscaling_zookeeper_root(service, instance)
-    with ZookeeperPool() as zookeeper_client:
-        zookeeper_client.ensure_path(zookeeper_path)
-        zookeeper_client.set(zookeeper_path, str(instance_count))
+class IngesterNoDataError(ValueError):
+    pass
 
 
-def get_instances_from_zookeeper(service, instance):
-    with ZookeeperPool() as zookeeper_client:
-        (instances, _) = zookeeper_client.get('%s/instances' % compose_autoscaling_zookeeper_root(service, instance))
-        return int(instances)
+@register_autoscaling_component('threshold', 'decider')
+def threshold_decider(marathon_service_config, ingester_method, marathon_tasks, mesos_tasks,
+                      delay=600, setpoint=0.8, threshold=0.1, **kwargs):
+    zk_last_time_path = '%s/threshold_last_time' % compose_autoscaling_zookeeper_root(
+        service=marathon_service_config.service,
+        instance=marathon_service_config.instance,
+    )
+    with ZookeeperPool() as zk:
+        try:
+            last_time, _ = zk.get(zk_last_time_path)
+            last_time = float(last_time)
+        except NoNodeError:
+            last_time = 0.0
 
+    current_time = int(datetime.now().strftime('%s'))
+    if current_time - last_time < delay:
+        return 0
 
-@register_autoscaling_method('bespoke')
-def bespoke_autoscaling_method(*args, **kwargs):
-    # do nothing, the service author has written their own scaling code
-    return 0
+    error = ingester_method(marathon_service_config, marathon_tasks, mesos_tasks, **kwargs) - setpoint
+
+    with ZookeeperPool() as zk:
+        zk.ensure_path(zk_last_time_path)
+        zk.set(zk_last_time_path, str(current_time))
+
+    if error > threshold:
+        return 1
+    elif abs(error) > threshold:
+        return -1
+    else:
+        return 0
+
+
+@register_autoscaling_component('pid', 'decider')
+def pid_decider(marathon_service_config, ingester_method, marathon_tasks, mesos_tasks,
+                delay=600, setpoint=0.8, **kwargs):
+    Kp = 0.2
+    Ki = 0.2 / delay
+    Kd = 0.05 * delay
+
+    autoscaling_root = compose_autoscaling_zookeeper_root(
+        service=marathon_service_config.service,
+        instance=marathon_service_config.instance,
+    )
+    zk_iterm_path = '%s/pid_iterm' % autoscaling_root
+    zk_last_error_path = '%s/pid_last_error' % autoscaling_root
+    zk_last_time_path = '%s/pid_last_time' % autoscaling_root
+
+    def clamp_value(number):
```

Comment:
what's the purpose of nested def here? it'll create a closure every time you call clamp_value and it doesn't seem like you need it.